File: day02/solve.R
#!/usr/bin/env Rscript
suppressPackageStartupMessages(library(tidyverse))

# Prep/process the input file into nice columns
dat <- "input.txt" %>%
    read_delim(
        delim = " ",
        col_names = c("min_max", "letter", "pass"),
        col_types = "ccc"
    ) %>%
    separate("min_max", c("min_n", "max_n"), sep = "-", convert = TRUE) %>%
    mutate(letter = str_remove(letter, ":.*$")) %>%
    rowwise()

print("Part 1 solution:")
dat %>%
    filter(between(str_count(pass, letter), min_n, max_n)) %>%
    nrow()

print("Part 2 solution:")
dat %>%
    # TRUE is 1, FALSE is 0; Only want one TRUE so sum of two
    # conditionals needs to be 1
    filter(
        1 == sum(
            letter == str_sub(pass, min_n, min_n),
            letter == str_sub(pass, max_n, max_n)
        )
    ) %>%
    nrow()

File: day02/solve.py
#!/usr/bin/env python3
import re

# lines formatted min-max letter: password
pattern = re.compile(r"^\s*(\d+)\-(\d+)\s+([a-zA-Z]):\s+([a-zA-Z]+)\s*$")

with open("input.txt", "r") as lines:
    passwords = [m.groups() for l in lines if (m := pattern.search(l)) is not None]

print(
    "Part 1 solution:",
    len([p for p in passwords if int(p[0]) <= p[3].count(p[2]) <= int(p[1])])
)

print(
    "Part 2 solution pt 2",
    sum((p[int(a) - 1] == l) ^ (p[int(b) - 1] == l) for a, b, l, p in passwords)
)
